terminate/terminate.go

```diff
@@ -0,0 +1,209 @@
+package terminate
+
+import (
+	"context"
+	"log"
+	"math"
+	"math/big"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-bitfield"
+	lotusapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	miner8 "github.com/filecoin-project/specs-actors/v8/actors/builtin/miner"
+
+	"github.com/glifio/go-pools/mstat"
+	"github.com/glifio/go-pools/util"
+	"golang.org/x/xerrors"
+)
+
+var PERC_SECTORS_TO_SAMPLE = big.NewRat(2, 100)
+var MAX_SAMPLED_SECTORS = big.NewInt(1000)
+var MIN_SAMPLED_SECTORS = big.NewInt(1)
+
+func EstimateTerminationPenalty(ctx context.Context, api *lotusapi.FullNodeStruct, minerAddr address.Address, ts *types.TipSet) (*big.Int, *big.Int, error) {
+	log.Println("Fetching all sectors ")
+	sectors, err := AllSectors(ctx, api, minerAddr, ts)
+	if err != nil {
+		return nil, nil, err
+	}
+	log.Println("Fetched all sectors ", len(sectors))
+
+	sampled := SampleSectors(sectors, PERC_SECTORS_TO_SAMPLE, MAX_SAMPLED_SECTORS, MIN_SAMPLED_SECTORS)
+
+	log.Println("Sampled sectors length: ", len(sampled))
+
+	// Sample 1% of sectors, with a max of 10000 and a minimum of 1000
+	sample := bitfield.NewFromSet(sampled)
+
+	res, err := TerminateSectors(ctx, api, minerAddr, &sample, ts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Printf("Termination penalty: %0.02f\n", util.ToFIL(res.TerminationFeeFromSample))
+
+	return res.EstimatedTotalTerminationFee, res.AvgTerminationFeePerPledge, nil
+}
+
+func AllSectors(ctx context.Context, api *lotusapi.FullNodeStruct, minerAddr address.Address, ts *types.TipSet) ([]uint64, error) {
+	allLiveSectors := make([]bitfield.BitField, 0)
+
+	for dlIdx := uint64(0); dlIdx < 48; dlIdx++ {
+		// 48 calls to the internet
+		partitions, err := api.StateMinerPartitions(ctx, minerAddr, uint64(dlIdx), ts.Key())
+		// if this error occurs, the miner hasn't been around long enough to process the deadlineTs, so we just skip it
+		if err != nil {
+			continue
+		}
+
+		for _, partition := range partitions {
+			allLiveSectors = append(allLiveSectors, partition.LiveSectors)
+		}
+	}
+
+	// one bitfield with all the sectors
+	allSectors, err := bitfield.MultiMerge(allLiveSectors...)
+	if err != nil {
+		return nil, err
+	}
+	// sectors => all sectors as an array of uint64 sector IDs
+	return allSectors.All(math.MaxUint64)
+}
+
+func SampleSectors(sectors []uint64, perc *big.Rat, max *big.Int, min *big.Int) []uint64 {
+	sectorCount := len(sectors)
+	if sectorCount == 0 {
+		return []uint64{}
+	}
+
+	// Compute the percentage of sectors
+	percFloat, _ := perc.Float64()
+	desiredCount := int(percFloat * float64(sectorCount))
```

Comment:
here we could simplify things and get rid of the `perc` and `min` logic - the `desiredCount` can start out as equal to `sectorCount`, and then just enforce the maximum. we just cap it